utils/sites.js

```diff
@@ -24,60 +23,121 @@ const submitParticipantsData = async (req, res, site) => {
-                let newStateElements = {}
-                for(let key in obj) {
-                    if(record.data.state[key] === undefined){
-                        newStateElements[`state.${key}`] = obj[key];
+
+    //let errorMsgs = [];
+
+    for(let dataObj of dataArray){
+        if(dataObj.token === undefined) {
+            error = true;
+            responseArray.push({'Invalid Request': {'Token': 'UNDEFINED', 'Errors': 'Token not defined in data object.'}});
+            continue;
+        }
+        
+        const participantToken = dataObj.token;
+        const { getParticipantData } = require('./firestore');
+        const record = await getParticipantData(participantToken, siteCode);
+
+        if(!record) {
+            error = true;
+            responseArray.push({'Invalid Request': {'Token': participantToken, 'Errors': 'Token does not exist.'}});
+            continue;
+        }
+
+        const docID = record.id;
+        const docData = record.data;
+
+        const flat = (obj, att, attribute) => {
+            for(let k in obj) {
+                if(typeof(obj[k]) === 'object') flat(obj[k], att, attribute ? `${attribute}.${k}`: k)
+                else flattened[att][attribute ? `${attribute}.${k}`: k] = obj[k]
+            }
+        }
+
+        let flattened = {
+            docData: {}
+        };
+
+        flat(docData, 'docData');
+
+        let errors = [];
+        
+        for(let key in dataObj) {
+
+            if(key == 'token') continue;
+            
+            if(flattened.docData[key]) {
+                errors.push(" Key (" + key + ") cannot exist before updating");
+                continue;
+            }
+
+            if(submitRules[key]) {
+                if(submitRules[key].dataType) {
+                    if(submitRules[key].dataType == 'ISO') {
+                        if(typeof dataObj[key] !== "string" || !(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/.test(dataObj[key]))) {
+                            errors.push(" Invalid data type / format for Key (" + key + ")");
+                        }
                     }
-                    // Make participant active if Study invitaion is sent.
-                    if(key === '934298480' && record.data['512820379'] !== 854703046) { // If age deidentified data is provided and participant is not passive then make this participant Active
-                        newStateElements['512820379'] = 486306141;
-                        newStateElements['471593703'] = new Date().toISOString();
+                    else {
+                        if(submitRules[key].dataType !== typeof dataObj[key]) {
+                            errors.push(" Invalid data type for Key (" + key + ")");
+                        }
+                        else {
+                            if(submitRules[key].values) {
+                                if(submitRules[key].values.filter(value => value.toString() === dataObj[key].toString()).length == 0) {
```

Comment:
`if (!submitRules[key].values.includes(dataObj[key]))`

Reply:
I will try this out, thanks.